relay-sampling/src/lib.rs

```diff
@@ -381,8 +427,58 @@ impl SamplingRule {
     ///
     /// Non-decaying rules are always active. Decaying rules are active if they are
     /// neither idle nor expired.
-    fn is_active(&self) -> bool {
-        self.time_range.contains(Utc::now())
+    fn is_active(&self, now: DateTime<Utc>) -> bool {
+        self.time_range.contains(now)
+    }
+
+    /// Returns the sample rate of this specific sampling rule based on the existence of the
+    /// decaying function.
+    ///
+    /// In case of any problems that render the decaying function impossible to compute, we are
+    /// going to fallback to the base sample rate.
+    pub fn get_sample_rate(&self, now: DateTime<Utc>) -> f64 {
+        if let Some(params) = self.get_decaying_function_params(now) {
+            self.decaying_fn.call(params)
+        } else {
+            self.sample_rate
```

Comment:
In my mind, if you pass a `timeRange` without `decayingFn`, you will use directly/implicitly the `Constant` function which only uses the `sampleRate` and the rule by itself is valid.

Reply:
@jan-auer this is a good point, with the new implementation, we would need to decide what to do in case of invalid params. We could either:
1. Implement a mechanism that automatically falls back either to `decayedSampleRate` or `sampleRate` based on the function type.
2. Make automatically the rule invalid by performing a validation of params before `get_sample_rate`
3. Propagate some error but this is generally not suggested and looking at the current implementation it is not done

_Just to clarify, this can happen for example if `timeRange` has `start` and not `end`, irrespectively if we specify a decaying function or not._

Reply:
We could write a custom deserializer for `SamplingConfig::rules`, which removes structurally invalid rules and logs an error for them. But `get_decaying_function_params` also depends on `now`, and that part of the "validation" should go into `is_active` IMO.

Reply:
I think that first we would need to clarify whether we want to allow open intervals or not.

If we wanted to do it better, I would have redefined the timerange with only closed intervals and then automatically use the constant decaying if not specified. Then for each function build custom extra validation and fallback mechanism(s).

I don't know tho, if there is a proper rationale behind the decision of open intervals. For DS we don't need it (as of now).